[ADD] website_sale_product_pack: New module

chienandalu opened the pull request:
Compatibility module between sale_product_pack and website_sale

cc @Tecnativa TT30385

chienandalu committed:
[ADD] website_sale_product_pack: New module

Compatibility module between sale_product_pack and website_sale

TT30385

pedrobaeza commented on website_sale_product_pack/controllers/main.py:
```diff
@@ -0,0 +1,13 @@
+    @http.route()
+    def product(self, product, category="", search="", **kwargs):
+        res = super().product(product, category=category, search=search, **kwargs)
+        http.request.env.context = dict(http.request.env.context, whole_pack_price=True)
```
Please add docstring about this. I think there's a way of not doing Python hack (maybe through `qcontext`?)

chienandalu replied:
It didn't work as expected anyway. I'm removing it

pedrobaeza commented on website_sale_product_pack/models/sale_order.py:
```diff
@@ -0,0 +1,37 @@
+        return super()._cart_update(product_id, line_id, add_qty, set_qty, **kwargs)
+
+    def _website_product_id_change(self, order_id, product_id, qty=0):
+        """In the final checkout stop, we could miss the component discount as the
```
```suggestion
        """In the final checkout step, we could miss the component discount as the
```

chienandalu replied:
fixed. thanks :)

pedrobaeza commented on website_sale_product_pack/models/sale_order.py:
```diff
@@ -0,0 +1,37 @@
+    def unlink(self):
+        """The website calls this method specifically. We want to get rid of
+        the children lines so the user doesn't have to"""
+        for line in self.filtered("pack_child_line_ids"):
+            self |= line.pack_child_line_ids
```
Don't add over `self`, but have a new variable (`obj` for example), and call super with it. You can as well have it directly:
```suggestion
        obj = self + self.mapped("pack_child_line_ids")
```

chienandalu replied:
fixed :+1:

chienandalu committed:
[FIX] stock_product_pack: get already computed values if existing

Done for e-commerce compatibility purposes, althoug it's more performant
indeed.

pedrobaeza commented on website_sale_product_pack/__manifest__.py:
```diff
@@ -0,0 +1,14 @@
+    "summary": "Compatibility module of product pack with e-commerce",
+    "version": "13.0.1.0.0",
+    "license": "AGPL-3",
+    "depends": ["website_sale", "sale_product_pack"],
```
Shouldn't this depend on `website_sale_stock`?

chienandalu replied:
A pack of services would be valid as well

pedrobaeza commented on website_sale_product_pack/views/assets.xml:
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <template id="assets_frontend" inherit_id="website_sale.assets_frontend">
```
Isn't there specific assets for tests? Or at least load it on demo.

chienandalu replied:
Yes. Done

chienandalu commented:
@ernestotejeda could you review?

pedrobaeza commented:
Ey, Travis goes to red now

chienandalu commented:
:green_apple:

nicomacr commented:
@chienandalu great work with this module. Really thanks
We got a similar module [here](https://github.com/ingadhoc/website/tree/13.0/website_product_pack) 

I've check this PR and compare with our module and find that would be interesting to have this features in the module to handle the publish mechanism:

- [here](https://github.com/ingadhoc/website/blob/13.0/website_product_pack/models/product_product.py#L13) we check for the products to add in a pack that all of them were published y if it not we alert. **Maybe with a domain in the field could be better.**

- [here](https://github.com/ingadhoc/website/blob/13.0/website_product_pack/models/product_template.py#L12) we set for the pack lines in a pack to be publish or not from the parent pack. And the inverse case if one of the pack component becomes to unpublish we set the same for the parent pack. **I think this part is better made in another method calling from write method.**
 
I thinks this could be necessary because the mixing of a product pack published with somes of the components not, could be confuse and some bug related to this will be happened.


Any thoughts about that?


Thanks again.

chienandalu commented:
Thanks @nicomacr I propose you to add your review in order to merge this one, and then I'll gladly review your PR with those improvements :slightly_smiling_face:

nicomacr reviewed (approved):
LGTM

pedrobaeza commented:
/ocabot merge nobump

pedrobaeza commented:
/ocabot merge nobump

pedrobaeza commented:
/ocabot merge nobump

pedrobaeza commented:
/ocabot merge nobump

pedrobaeza commented:
/ocabot merge nobump

pedrobaeza commented:
@chienandalu rebase to see the problem.

alvaro-domatix commented:
We've detected a problem trying to use the packs on the website.

If you configure a product as a pack with pack_type = "detailed" and pack_component_price = "ignored" it apparently works while you do the cart until you try to pay. Once you enter the address information and click on "Next", the order sums the amount of all the components and the pack too.

Example:

Pack 1 (30€):
- Comp 1 (20€)
- Comp 2 (20€)
Once you click on next after the address information the order's amount is 30€ + 20€ + 20€ = 70€

ernestotejeda commented on website_sale_product_pack/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,10 @@
+- The cart popup summary only shows the main pack line and discards the sublines in
+  the units count.
+- The cart summary shows the component lines hanging from the main one as well.
+- It's ensured the the prices are shown correctly for the whole pack and that they're
```
```suggestion
- It's ensured the prices are shown correctly for the whole pack and that they're
```

ernestotejeda reviewed (commented):
I have functionally tested the module and I have found the same error that @alvaro-domatix  mentions here https://github.com/OCA/product-pack/pull/46#issuecomment-926829089
Also, when a pack is removed from the cart, its components are not removed.

dreispt commented:
/ocabot merge nobump
